### PR TITLE
Allow non blocking read from TabNine process

### PR DESCRIPTION
**Pull request by tzachar**

### CODE REVIEW

**Comment by Shougo** [comment]
Thanks.  I use it for a while.

**Comment by Shougo** [comment]
@tbodt Can you check the implementation?

**Comment by Shougo** [comment]
The error exists.

```diff
diff --git a/rplugin/python3/deoplete/sources/tabnine.py b/rplugin/python3/deoplete/sources/tabnine.py
index 3ac0f52..396f3a4 100644
--- a/rplugin/python3/deoplete/sources/tabnine.py
+++ b/rplugin/python3/deoplete/sources/tabnine.py
@@ -154,7 +154,7 @@ class Source(Base):
             output = rqueue.get(block=True, timeout=1)
             return json.loads(output)
         except queue.Empty:
-            self.debug('Tabnine output is corrupted: ' + output)
+            return
         except json.JSONDecodeError:
             self.debug('Tabnine output is corrupted: ' + output)
```

`output` is not assigned in the except.

**Comment by tzachar** [comment]
Yeah, also noticed it. Its a small fix but is irrelevant to the main issue.

File: rplugin/python3/deoplete/sources/tabnine.py [hunk 1]
```diff
@@ -144,16 +150,20 @@ def _request(self, name, **params):
             self._restart()
             return
 
-        output = proc.stdout.readline()
         try:
+            output = rqueue.get(block=True, timeout=1)
```
Comment (Shougo) [inline rplugin/python3/deoplete/sources/tabnine.py]: I think `timeout=0.5` is better for me.

**Comment by tbodt** [comment]
Is the goal to have a timeout on the read call? If so, it would be simpler to use the [`selectors`](https://docs.python.org/3/library/selectors.html) module and nonblocking I/O to do that directly.

**Comment by Shougo** [comment]
> The type of file objects supported depends on the platform: on Windows, sockets are supported, but not pipes, whereas on Unix, both are supported (some other types may be supported as well, such as fifos or special file devices).

It does not support Windows OK?

**Comment by tzachar** [comment]
Is Windows a major concern?

**Comment by Shougo** [comment]
I don't use it in Windows though

**Comment by tzachar** [comment]
Added a simple commit which now properly closes the readQ when the process dies.
Seems to work better.

Also lowered the read timeout to 0.5 seconds like @Shougo suggested.

**Comment by Shougo** [comment]
Thanks. I will test it.

**Comment by Shougo** [comment]
Hm.... Your changes freezes my neovim.
I don't know why but neovim CPU usage is very high sometimes.

Note: If I use original code, the behavior is not occur.

**Comment by Shougo** [comment]
Using `selectors` may be better than current implementation.

**Comment by tzachar** [comment]
Strange. It actaully improved my experience.

**Comment by tzachar** [comment]
`selectors` are definately better. 

Wonder when someone will make the effort to implement an open source alternative to TabNine....

**Comment by Shougo** [comment]
> Wonder when someone will make the effort to implement an open source alternative to TabNine....

Yeah.

**Comment by Shougo** [comment]
I think ` rqueue.get(block=True, timeout=1)` causes the freeze.
Non blocking implementation is better.

**Comment by tzachar** [comment]
Moved the patch to use selectors. Try now.

**Comment by Shougo** [comment]
スクリーンショット_2020-11-02_19-08-35

The error is occurred.

**Comment by tzachar** [comment]
Sorry, my bad. removed the dead code now.

**Comment by Shougo** [comment]
OK.

**Comment by Shougo** [comment]
The implementation seems better than the previous version.
I use it take a while.

More testers are needed.

**Comment by tzachar** [comment]
I've been using it for a while now, and it seems to make things much better.
nvim used to hang on exit, which did not happen with the patch.
Moreover, the last addition of the suicide option also solved the excessive TabNine memory consumption problem.

**Comment by Shougo** [comment]
It seems better and no problem now.

File: rplugin/python3/deoplete/sources/tabnine.py [hunk 2]
```diff
@@ -154,19 +162,28 @@ def _restart(self):
...
         path = get_tabnine_path(binary_dir)
         if path is None:
             self.print_error('no TabNine binary found')
             return
         self._proc = subprocess.Popen(
-            [path, '--client', 'sublime', '--log-file-path',
-             os.path.join(self._install_dir, 'tabnine.log')],
+            [
+                path,
+                '--client', 'sublime',
```
Comment (boaz-codota) [inline rplugin/python3/deoplete/sources/tabnine.py]: Could you also change the `--client` flag to be `deoplete`?

File: rplugin/python3/deoplete/sources/tabnine.py [hunk 3]
```diff
@@ -144,8 +145,15 @@ def _request(self, name, **params):
             self._restart()
             return
 
-        output = proc.stdout.readline()
         try:
+            # try to read from proc's stdout
+            events = selector.select(timeout=0.5)
```
Comment (boaz-codota) [inline rplugin/python3/deoplete/sources/tabnine.py]: We advise for 1 second timeout, as we've also implemented in vscode & intellij plugins:
https://github.com/codota/tabnine-intellij/blob/b8ddb130ec007a757e6fc181a9bc745cf946bac7/src/main/java/com/tabnine/binary/BinaryRequestFacade.java#L26

File: rplugin/python3/deoplete/sources/tabnine.py [hunk 4]
```diff
@@ -144,8 +145,15 @@ def _request(self, name, **params):
...
+            if len(events) == 0:
+                # nothing from TabNine. Restart it
+                self._restart()
```
Comment (boaz-codota) [inline rplugin/python3/deoplete/sources/tabnine.py]: This is rather aggressive, TabNine tends to not return results in when the prediction is for a newline. So this we will cause way more restarts than needed and will damage the experience.

I would also like to point you to our implementation for that in IntelliJ:
https://github.com/codota/tabnine-intellij/blob/b8ddb130ec007a757e6fc181a9bc745cf946bac7/src/main/java/com/tabnine/binary/TabNineGateway.java#L98

Here we validate results based on schema, and only upon a few consecutive bad results we restart the process.

File: rplugin/python3/deoplete/sources/tabnine.py [hunk 5]
```diff
@@ -154,19 +162,28 @@ def _restart(self):
...
             self.print_error('no TabNine binary found')
             return
         self._proc = subprocess.Popen(
-            [path, '--client', 'sublime', '--log-file-path',
-             os.path.join(self._install_dir, 'tabnine.log')],
+            [
+                path,
+                '--client', 'sublime',
+                '--log-file-path', os.path.join(self._install_dir, 'tabnine.log'),
+                '--idle_suicide_seconds', '3600',
```
Comment (boaz-codota) [inline rplugin/python3/deoplete/sources/tabnine.py:188]: Unfortunately, this is currently has no effect on the operation of the binary.

**Review by boaz-codota — COMMENTED** [review]
Thanks for doing this PR. Timeouts are an awesome idea that we've also started implementing a while ago in our internally mentained plugins.

I just have a few suggestions to make this hopefully even better.

File: rplugin/python3/deoplete/sources/tabnine.py (hunk 2 above)
Comment (tzachar) [reply rplugin/python3/deoplete/sources/tabnine.py]: Does it make any difference in the operation of TabNine?

File: rplugin/python3/deoplete/sources/tabnine.py (hunk 3 above)
Comment (tzachar) [reply rplugin/python3/deoplete/sources/tabnine.py]: @Shougo reported better behaviour with 0.5 seconds. Was also my experince.

File: rplugin/python3/deoplete/sources/tabnine.py (hunk 4 above)
Comment (tzachar) [reply rplugin/python3/deoplete/sources/tabnine.py]: As far as I understand, deoplete does not request a completion for an empty line, unless specifically configured to do so.

File: rplugin/python3/deoplete/sources/tabnine.py (hunk 5 above)
Comment (tzachar) [reply rplugin/python3/deoplete/sources/tabnine.py:188]: Thats a shame, as the binary tends to bloat over time. Last week it gorged itself on all my RAM, all 128 GB of it.

File: rplugin/python3/deoplete/sources/tabnine.py (hunk 4 above)
Comment (tzachar) [reply rplugin/python3/deoplete/sources/tabnine.py]: Another question; how do you keep the process of asking for completions and getting replies in sync if you ignore a bad respons?

File: rplugin/python3/deoplete/sources/tabnine.py (hunk 3 above)
Comment (Shougo) [reply rplugin/python3/deoplete/sources/tabnine.py]: `0.5` is for block timeout.  It is not block timeout.  So I think to use `1.0` is acceptable.

File: rplugin/python3/deoplete/sources/tabnine.py (hunk 2 above)
Comment (tbodt) [reply rplugin/python3/deoplete/sources/tabnine.py]: In any case this should be a different PR

**Comment by Shougo** [comment]
TabNine 3.2.2 seems released now.

**Comment by tzachar** [comment]
@Shougo , I've been using this patch for the last couple of days, and it seems to be doing a pretty good job of isolating neovim from TabNine's tantrums. What about a merge?

**Comment by Shougo** [comment]
I have been used it and no problem.
I think the PR is ready to merge.

@tbodt Can you merge it?

**Comment by tzachar** [comment]
No progress..... 
Maybe we should fork?

**Comment by tbodt** [comment]
Sorry, missed this message.

I'm concerned about windows support: this plugin does support windows, and this PR would likely break it. Can we have a fallback to the old code for this scenario?

**Comment by tzachar** [comment]
I don't have a working Windows setup to test it.

**Comment by Shougo** [comment]
You can add Win32 check like this.  I think the code is better.

```diff
diff --git a/rplugin/python3/deoplete/sources/tabnine.py b/rplugin/python3/deoplete/sources/tabnine.py
index 2a44598..53ec7b8 100644
--- a/rplugin/python3/deoplete/sources/tabnine.py
+++ b/rplugin/python3/deoplete/sources/tabnine.py
@@ -4,6 +4,7 @@ import os
 import platform
 import subprocess
 import selectors
+import sys

 from deoplete.source.base import Base
 from deoplete.util import getlines
@@ -146,12 +147,14 @@ class Source(Base):
             return

         try:
-            # try to read from proc's stdout
-            events = selector.select(timeout=1.0)
-            if len(events) == 0:
-                # nothing from TabNine. Restart it
-                self._restart()
-                return
+            if self._selector is not None:
+                # try to read from proc's stdout
+                events = selector.select(timeout=1.0)
+                if len(events) == 0:
+                    # nothing from TabNine. Restart it
+                    self._restart()
+                    return
+
             # ok, we can read. we assume there is a single file attached to the selector.
             output = proc.stdout.readline()
             return json.loads(output)
@@ -162,8 +165,11 @@ class Source(Base):
         if self._proc is not None:
             self._proc.terminate()
             self._proc = None
+
+        if self._selector is not None:
             self._selector.close()
             self._selector = None
+
         binary_dir = os.path.join(self._install_dir, 'binaries')
         path = get_tabnine_path(binary_dir)
         if path is None:
@@ -182,8 +188,10 @@ class Source(Base):
             encoding='utf-8',
             close_fds=True,
         )
-        self._selector = selectors.DefaultSelector()
-        self._selector.register(self._proc.stdout, selectors.EVENT_READ)
+        if sys.platform != 'win32':
+            # Note: Windows doesn't support pipe selector.
+            self._selector = selectors.DefaultSelector()
+            self._selector.register(self._proc.stdout, selectors.EVENT_READ)

     def _get_running_tabnine(self):
         if self._proc is None:
```

File: rplugin/python3/deoplete/sources/tabnine.py (hunk 5 above)
Comment (boazberman) [reply rplugin/python3/deoplete/sources/tabnine.py:188]: We have code inside the binary that should check such things. If you come across such bug again please open us an issue or contact support. We would love to resolve it!

File: rplugin/python3/deoplete/sources/tabnine.py (hunk 4 above)
Comment (boazberman) [reply rplugin/python3/deoplete/sources/tabnine.py]: If there is a bas result (exception level bad) we restart the binary. If the response is just not parsing correctly or returns as null we try again (up to 5 consecutive times, on the 5th time we restart).
Please consult the tabnine-intellij plugin and vscode plugin for implementation details.

P. S. Sorry for the late response, my email is bloated AF.

File: rplugin/python3/deoplete/sources/tabnine.py (hunk 2 above)
Comment (boazberman) [reply rplugin/python3/deoplete/sources/tabnine.py]: It helps us to spot bugs and their origin

File: rplugin/python3/deoplete/sources/tabnine.py (hunk 4 above)
Comment (tzachar) [reply rplugin/python3/deoplete/sources/tabnine.py]: The solution as it currently stands is simpler. Do you have any statistics on the average number of tries needed in the intelij plugin? I would bet it's heavily skewed towards working on the first time or failing all 5 attempts. If so, I don't see the need to complicate the code when a simple restart works.

Comment (boaz-codota) [reply rplugin/python3/deoplete/sources/tabnine.py]: Yep your are right, but 0 results is a valid result. `null` is also a valid result in some cases...

Comment (tzachar) [reply rplugin/python3/deoplete/sources/tabnine.py]: The problem is we need to distinguish between an empty reply and the daemon dying on us or taking too long to reply.
The current approach works. Why change it?

Comment (boaz-codota) [reply rplugin/python3/deoplete/sources/tabnine.py]: Your call, I explained the constraints. It sure can work, it just do much more restarts than needed, that might impact the experience.

**Comment by Shougo** [comment]
LGTM

@tbodt Ping.

**Comment by Shougo** [comment]
OH, I have found the bug.

The initialization is needed.

```diff
diff --git a/rplugin/python3/deoplete/sources/tabnine.py b/rplugin/python3/deoplete/sources/tabnine.py
index 6d8df5c..652b7c6 100644
--- a/rplugin/python3/deoplete/sources/tabnine.py
+++ b/rplugin/python3/deoplete/sources/tabnine.py
@@ -61,6 +61,7 @@ class Source(Base):

         self._proc = None
         self._response = None
+        self._selector = None
         self._install_dir = os.path.dirname(os.path.dirname(os.path.dirname(
             os.path.dirname(os.path.dirname(os.path.realpath(__file__))))))

```

**Comment by tzachar** [comment]
> OH, I have found the bug.
> 
> The initialization is needed.
> 
> ```diff
> diff --git a/rplugin/python3/deoplete/sources/tabnine.py b/rplugin/python3/deoplete/sources/tabnine.py
> index 6d8df5c..652b7c6 100644
> --- a/rplugin/python3/deoplete/sources/tabnine.py
> +++ b/rplugin/python3/deoplete/sources/tabnine.py
> @@ -61,6 +61,7 @@ class Source(Base):
> 
>          self._proc = None
>          self._response = None
> +        self._selector = None
>          self._install_dir = os.path.dirname(os.path.dirname(os.path.dirname(
>              os.path.dirname(os.path.dirname(os.path.realpath(__file__))))))
> ```

Fixed in latest commit.

**Comment by tbodt** [comment]
Looks good to me now, if no complaints I'll merge tomorrow.

**Comment by tbodt** [comment]
Thanks for the PR!

**Pull request merged**